src/tool/subcommands/car_cmd.rs

```diff
@@ -50,7 +69,166 @@ impl CarCommands {
                 crate::db::car::forest::Encoder::write(&mut writer, all_roots, frames).await?;
                 writer.flush().await?;
             }
+            Self::Validate {
+                car_file,
+                ignore_block_validity,
+                ignore_forest_index,
+            } => validate(&car_file, ignore_block_validity, ignore_forest_index).await?,
         }
         Ok(())
     }
 }
+
+/// At present, three properties are checked:
+/// - The CAR file is syntactically valid and all blocks can be streamed.
+/// - Each block CID is checked against the hash of the block.
+/// - Each block CID is looked-up in the on-disk index.
+///
+/// Properties related to Filecoin are not checked. For those, see `forest-tool
+/// snapshot validate`.
+///
+/// We do not check for duplicate blocks. Whether duplicate blocks are allowed or
+/// not is vague in the specification.
+async fn validate(
+    car_file: &Path,
+    ignore_block_validity: bool,
+    ignore_forest_index: bool,
+) -> anyhow::Result<()> {
+    let optional_db = if !ignore_forest_index {
+        Some(ForestCar::try_from(car_file)?)
+    } else {
+        None
+    };
+
+    let file = File::open(car_file).await?;
+    let pb = ProgressBar::new(file.metadata().await?.len()).with_style(
+        ProgressStyle::with_template("{bar} {percent}%, eta: {eta}").expect("infallible"),
+    );
+    let file = BufReader::new(pb.wrap_async_read(file));
+
+    let mut stream = CarStream::new(file).await?;
+    while let Some(block) = stream.try_next().await? {
+        if !ignore_block_validity && !block.valid() {
+            anyhow::ensure!(block.valid(), "CID/Block mismatch for block: {}", block.cid);
+        }
+        if let Some(ref db) = optional_db {
+            anyhow::ensure!(db.get(&block.cid).ok().flatten() == Some(block.data));
+        }
+    }
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::validate;
+    use crate::db::car::forest;
+    use crate::networks::{calibnet, mainnet};
+    use crate::utils::db::car_stream::CarBlock;
+    use cid::multihash::{Code, MultihashDigest};
+    use cid::Cid;
+    use futures::{stream::iter, StreamExt, TryStreamExt};
+    use std::io::Write;
+    use tempfile::{NamedTempFile, TempPath};
+    use tokio::io::AsyncWriteExt;
+
+    #[tokio::test]
+    async fn validate_junk_car() {
+        let mut temp_path = NamedTempFile::new_in(".").unwrap();
```

Comment:
Why does it need to be in `.`? In general, I'd prefer that we don't create test artefacts in the source code directory, even temporarily.

Reply:
It doesn't. But you banned `NamedTempFile::new`.

Reply:
How about `tempfile::Builder::new().tempfile().unwrap()`?

Reply:
Done.